### PR TITLE
Handle PrivateKeyProvider configuration

### PR DESCRIPTION
**Pull request by rveerama1**
**Please provide a description of this PR:**

When the PrivateKeyProvider configuration information is provided through Istio operator yaml file, parse and pass it to gateway or sidecars. Envoy will act based on the information provided by the configuration.

To set the mesh wide defaults, configure the `defaultConfig` section of `meshConfig`.
For example:

```
meshConfig:
  defaultConfig:
    privateKeyProvider:
      cryptomb:
        pollDelay: 0.01s
```

This can also be configured on a per-workload basis by configuring the `proxy.istio.io/config`
annotation on the pod.
For example:

```
annotations:
  proxy.istio.io/config: |
    privateKeyProvider:
      cryptomb:
        pollDelay: 0.01s
```

This requires https://github.com/istio/api/pull/2261

### CODE REVIEW

**Review by Arhell — COMMENTED** [review]
/ok-to-test

**Comment by rveerama1** [comment]
As it depends upon [api](https://github.com/istio/api/pull/2261) changes, build fails. I will move to WIP.

**Comment by rveerama1** [comment]
@howardjohn @jacob-delgado @ericvn PTAL about implementation of https://github.com/istio/api/pull/2261private key provider addition.

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (yangminzhu) [inline security/pkg/nodeagent/sds/sdsservice.go:275]: nit, a little bit more go idiom if `res` is not used beyond the if-statement:
```
if res := toEnvoySecretWithPrivateKeyProvider(s, pkpConf); res != nil {
 ...
} else {
 ...
}
```

File: pilot/pkg/xds/sds.go
Comment (yangminzhu) [inline pilot/pkg/xds/sds.go:312]: this has quite some duplicate code as the other `toEnvoyKeyCertSecretWithPrivateKeyProvider()`, could we refactor and try to share the common logic

File: security/pkg/nodeagent/sds/server.go
Comment (yangminzhu) [inline security/pkg/nodeagent/sds/server.go:47]: Just some general questions

1. Regarding the `PrivateKeyProvider` extension, is it actually providing any private key? It seems a bit confused to be because based on the PR and my understanding, it does not provide any private key but is just a new type (cryptomb) to wrap the TLS config that tells Envoy to  process it with some hardware assisted acceleration?

2. How are you going to e2e test the feature if it requires some special hardware support?

File: pilot/pkg/bootstrap/server.go
Comment (jacob-delgado) [inline pilot/pkg/bootstrap/server.go:518]: I think you may want to use `EffectiveProxyConfig` as there is a merge issue with meshConfig vs a proxy annotation.

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (jacob-delgado) [inline security/pkg/nodeagent/sds/sdsservice.go:186]: Do you need to pass in `s.pkgConf` as you are calling a receiver on `s` anyway?

File: security/pkg/nodeagent/sds/server.go
Comment (rveerama1) [reply security/pkg/nodeagent/sds/server.go:47]: > 1. Regarding the `PrivateKeyProvider` extension, is it actually providing any private key? The name is a bit confused to me because based on the PR and my understanding, it does not provide any private key but is just a new type (cryptomb) to wrap the TLS config that tells Envoy to  process it with some hardware assisted acceleration?

Name is bit confusing, but I tried to sync with Envoy naming pattern https://github.com/envoyproxy/envoy/tree/main/api/contrib/envoy/extensions/private_key_providers

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (rveerama1) [reply security/pkg/nodeagent/sds/sdsservice.go]: Done

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (rveerama1) [reply security/pkg/nodeagent/sds/sdsservice.go]: Done

File: pilot/pkg/xds/sds.go
Comment (rveerama1) [reply pilot/pkg/xds/sds.go:312]: It has some, but not completely. Anyway I am working on QAT private key provider extension support. I will fine tune in that PR.

File: security/pkg/nodeagent/sds/server.go
Comment (ipuustin) [reply security/pkg/nodeagent/sds/server.go:47]: I think there are two things which can be e2e tested. One is the functionality of setting a private key provider configuration to Envoy, and the other is a specific private key provider such as CryptoMB.

The first case is probably not too bad -- for example Envoy uses a software-only private key provider for testing the interfaces (https://github.com/envoyproxy/envoy/blob/main/test/extensions/transport_sockets/tls/test_private_key_method_provider.cc) and we could do something similar. The test private key provider isn't exposed by Envoy though, but that code could probably be used or reimplemented by istio/proxy test/envoye2e directory.

The second case (CryptoMB) is tricky because that will indeed require a very modern Xeon processor. An e2e test could be written but it should be behind a switch, and we need to understand if it's possible to make the CI tests run in an environment which supports AVX-512. @yangminzhu do you have any suggestions how this could be accomplished?

File: pilot/pkg/bootstrap/server.go
Comment (howardjohn) [reply pilot/pkg/bootstrap/server.go:518]: As far as I can tell this is not for Istiod, this is for responding to Gateway requests. Which still have per-node Proxyconfig we need to respect. So I agree with Jacob

File: pilot/pkg/xds/sds.go
Comment (howardjohn) [inline pilot/pkg/xds/sds.go:354]: style nit: I don't think we need these split code paths for `pkpConf != nil`. We can just have a big `switch` that has cases for `mesh.PrivateKeyProvider_Cryptomb`, <any future providers>, and `default`; the default case is the current code.

Essentially the default case *is* a PrivateKeyProvider

File: pkg/config/validation/validation.go
Comment (howardjohn) [inline pkg/config/validation/validation.go:1641]: isn't `Duration { seconds: 1, nanos: 0 }` valid?

File: pilot/pkg/bootstrap/server.go
Comment (rveerama1) [reply pilot/pkg/bootstrap/server.go:518]: I mean istiod to gateways. I will check how can I use EffectiveProxyconfig.

File: pkg/config/validation/validation.go
Comment (rveerama1) [reply pkg/config/validation/validation.go:1641]: Right, sorry about that, my wooden eyes did see it as oneof{seconds, nanos}. I will fix it.

**Comment by rveerama1** [comment]
@howardjohn @jacob-delgado updated the PR. Please take a look.

File: security/pkg/nodeagent/sds/server.go
Comment (yangminzhu) [reply security/pkg/nodeagent/sds/server.go:47]: Thanks for the detail explanation, the naming part is a bit confusing but I think it is fine if you could document it clearly in the code and documentation so that other users does not confuse it with things like Plug in certificate that is really providing its own private/public key to Istio.

Regarding the e2e tests, it would be nice if you can somehow mock the CryptoMB functionality in Envoy then run it like a real e2e test but I have no idea how to do it easily so you probably will end up with tests verifying the Envoy TLS configs generated correctly (but not executing them). Also I have no idea how to require a very modern Xeon processor in the CI environment (it gives you some flexibility in machine type but probably not to the level for custom CPU), @howardjohn might know about about the CI configuration.

File: pilot/pkg/bootstrap/server.go
Comment (rveerama1) [reply pilot/pkg/bootstrap/server.go]: Found a reference usage [here](https://github.com/istio/istio/blob/master/pkg/kube/inject/webhook.go#L876). Updated the PR.

**Comment by howardjohn** [comment]
How we do this with other custom integrations is OSS has unit level tests that check the generated XDS is correct. The owner of the platform runs tests on their own system with their own platform to verify it works in the real world.

For example we do this with things like Stackdriver, etc

**Comment by rveerama1** [comment]
> The owner of the platform runs tests on their own system with their own platform to verify it works in the real world.

Sounds a bit bigger task to me to prepare a such setup and tests :). But I will investigate and work on it.

Are the rest of the changes are fine?

File: pilot/pkg/xds/sds.go
Comment (howardjohn) [inline pilot/pkg/xds/sds.go:320]: @jacob-delgado  @Monkeyanator this doesn't seem to take `meta.ProxyConfig` into account which makes me suspicious that its not quite right. WDYT?

I mean the existing EffectiveProxyConfig code, not the new stuff

**Review by howardjohn — COMMENTED** [review]
Looks good except the one proxy config comment

File: pilot/pkg/xds/sds.go
Comment (rveerama1) [reply pilot/pkg/xds/sds.go:319]: Gentle ping @jacob-delgado @Monkeyanator

Comment (howardjohn) [reply pilot/pkg/xds/sds.go:319]: Maybe we should be using ProxyConfigOrDefault?

EffectiveProxyConfig is used only for injection

Comment (Monkeyanator) [reply pilot/pkg/xds/sds.go:319]: Yeah `ProxyConfigOrDefault` should be used here I think. `EffectiveProxyConfig` is used during injection which is where `meta.ProxyConfig` gets set from IIRC.

Comment (rveerama1) [reply pilot/pkg/xds/sds.go]: Thanks for the comments, I have now modified to use default.

Comment (rveerama1) [reply pilot/pkg/xds/sds.go]: PTAL @jacob-delgado

**Comment by rveerama1** [comment]
> Pod annotations may not be working.

@jacob-delgado I tried to modify httpbin sample with pod annotation without mentioning in global meshconfig and it worked with on pod with cryptomb hardware.

`apiVersion: apps/v1
kind: Deployment
metadata:
  name: httpbin
spec:
  replicas: 1
  selector:
    matchLabels:
      app: httpbin
      version: v1
  template:
    metadata:
      labels:
        app: httpbin
        version: v1
      annotations:
        proxy.istio.io/config: |
          privateKeyProvider:
            cryptomb:
              pollDelay: 170ms
    spec:
      serviceAccountName: httpbin
      containers:
      - image: docker.io/kennethreitz/httpbin
        imagePullPolicy: IfNotPresent
        name: httpbin
        ports:
        - containerPort: 80
`

File: pilot/pkg/xds/sds.go
Comment (howardjohn) [reply pilot/pkg/xds/sds.go:313]: Yeah i tested annotation only works with workload cert, not gateway cert, because of this.

**Review by howardjohn — APPROVED** [review]
LGTM once we fix the gateway part, should be small fix

File: pilot/pkg/xds/sds.go
Comment (rveerama1) [reply pilot/pkg/xds/sds.go]: Thank you, fixed now.

File: pilot/pkg/xds/sds.go
Comment (howardjohn) [inline pilot/pkg/xds/sds.go:314]: can you remove this? other lgtm

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/xds/sds.go
Comment (rveerama1) [reply pilot/pkg/xds/sds.go]: sorry about that.

**Review by jacob-delgado — APPROVED** [review]

**Comment by jacob-delgado** [comment]
cc @rveerama1 Looks like you need to re-base, but have the appropriate approvals.

**Review by howardjohn — COMMENTED** [review]
Thanks for the great work here!

**Pull request merged**
